uisdk/src/main/java/com/karhoo/uisdk/screen/booking/booking/payment/adyen/AdyenPaymentPresenter.kt

```diff
@@ -74,48 +96,82 @@ class AdyenPaymentPresenter(view: PaymentMVP.View,
-        this.sdkToken = sdkNonce
+    override fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        if (resultCode == AppCompatActivity.RESULT_OK && data != null) {
+            val dataString = data.getStringExtra(AdyenResultActivity.RESULT_KEY)
+            val payload = JSONObject(dataString)
+            when (payload.optString(AdyenPaymentView.RESULT_CODE, "")) {
+                AdyenPaymentView.AUTHORISED -> {
+                    val transactionId = payload.optString(AdyenPaymentView.MERCHANT_REFERENCE, "")
+                    updateCardDetails(nonce = transactionId,
+                                      paymentData = payload.optString
+                                      (AdyenPaymentView.ADDITIONAL_DATA, null))
+                    passBackNonce(transactionId)
+                }
+                else -> view?.showPaymentFailureDialog()
+            }
+        } else {
+            view?.showPaymentFailureDialog()
+        }
+    }
+
+    override fun initialiseGuestPayment(price: QuotePrice?) {
+        passBackThreeDSecureNonce(sdkToken, quotePriceToAmount(price))
     }
 
     override fun onSavedPaymentInfoChanged(userPaymentInfo: SavedPaymentInfo?) {
         view?.bindPaymentDetails(savedPaymentInfo = userPaymentInfo)
     }
 
+    override fun passBackNonce(sdkNonce: String) {
+        this.sdkToken = sdkNonce
+    }
+
+    private fun passBackThreeDSecureNonce(nonce: String, amount: String) {
+        view?.threeDSecureNonce(sdkToken, nonce, amount)
+    }
+
+    private fun quotePriceToAmount(price: QuotePrice?): String {
+        val currency = Currency.getInstance(price?.currencyCode?.trim())
+        return CurrencyUtils.intToPriceNoSymbol(currency, price?.highPrice.orZero())
+    }
+
+    override fun sdkInit(price: QuotePrice?) {
+        this.price = price
+        paymentsService.getAdyenPublicKey().execute { result ->
+            when (result) {
+                is Resource.Success -> {
+                    result.data.let {
+                        adyenKey = it.publicKey
+                        getPaymentMethods()
+                    }
+                }
+                //TODO Change error message
+                is Resource.Failure -> view?.showError(R.string.payment_issue_message)
+            }
+        }
+    }
+
+    override fun setSavedCardDetails() {
+        view?.bindPaymentDetails(userStore.savedPaymentInfo)
+    }
+
     override fun updateCardDetails(nonce: String, cardNumber: String?, typeLabel: String?,
                                    paymentData: String?) {
         this.nonce = nonce
         paymentData?.let {
             val additionalData = JSONObject(paymentData)
-            val cardNumber = additionalData.optString("cardSummary", "")
-            val type = additionalData.optString("paymentMethod", "")
-            //TODO Use CardType fromString
+            val cardNumber = additionalData.optString(CARD_SUMMARY, "")
```

Comment:
a bit confusing to have the same name as the function parameter. I understand the function parameter is not used in Adyen

Reply:
Not quite sure what you mean.....

Reply:
The local variable is called `val cardNumber` and the function parameter is called the same `override fun updateCardDetails(nonce: String, cardNumber: String?, typeLabel: String?, paymentData: String?) {`

Not a bit issue. but when reading the code if you miss the local variable declaration, you may assume it is the function parameter ....